db/src/tx.rs

```diff
@@ -749,28 +757,115 @@ fn conclude_tx(tx: Tx, report: TxReport) -> Result<(TxReport, PartitionMap, Opti
+pub fn transact_terms<'conn, 'a, 'id, I>(conn: &'conn rusqlite::Connection,
                                     partition_map: PartitionMap,
                                     schema_for_mutation: &'a Schema,
                                     schema: &'a Schema,
                                     terms: I,
-                                    tempid_set: InternSet<TempId>) -> Result<(TxReport, PartitionMap, Option<Schema>)>
+                                    tempid_set: InternSet<TempId>) -> Result<(TxReport, PartitionMap, Option<Schema>, AttributeSet)>
     where I: IntoIterator<Item=TermWithTempIds> {
-
     let mut tx = start_tx(conn, partition_map, schema_for_mutation, schema)?;
-    let report = tx.transact_simple_terms(terms, tempid_set)?;
-    conclude_tx(tx, report)
+    let (report, changes) = tx.transact_simple_terms(terms, tempid_set)?;
+    conclude_tx(tx, report, changes)
+}
+
+#[derive(Default, Debug, Clone)]
+pub struct BatchedTransaction {
+    transactions: BTreeMap<Entid, AttributeSet>
+}
+
+impl BatchedTransaction {
+    pub fn add_transact(&mut self, txid: Entid, changes: AttributeSet ) {
+        self.transactions.insert(txid, changes);
+    }
+
+    pub fn get(&self) -> &BTreeMap<Entid, AttributeSet> {
+        &self.transactions
+    }
+}
+
+pub struct TxObserver {
+    notify_fn: Option<Box<FnMut(String, BatchedTransaction)>>
+}
+
+impl TxObserver {
+    pub fn new<F>(notify_fn: F) -> TxObserver where F: FnMut(String, BatchedTransaction) + 'static {
+        TxObserver { notify_fn: Some(Box::new(notify_fn)) }
+    }
+
+    fn notify(&mut self, key: String, transactions: BatchedTransaction) {
+        if let Some(ref mut notify_fn) = self.notify_fn {
+            (notify_fn)(key, transactions);
+        } else {
+            eprintln!("no notify function specified for TxObserver");
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct TxObservationService {
+    observers: BTreeMap<String, TxObserver>,
+    attributes: BTreeMap<String, AttributeSet>,
+}
+
+impl<'o> TxObservationService {
+    // For testing purposes
+    pub fn is_registered(&self, key: &String) -> bool {
+        self.observers.contains_key(key)
+    }
+
+    pub fn register(&mut self, observer: TxObserver, key: String, attributes: AttributeSet) {
+        self.observers.insert(key.clone(), observer);
+        self.attributes.insert(key, attributes);
+    }
+
+    pub fn deregister(&mut self, key: &String) {
+        self.observers.remove(key);
+        self.attributes.remove(key);
+    }
+
+    fn observers_for_attributes(&self, changes: &AttributeSet) -> BTreeSet<(String, AttributeSet)> {
+        // filter observers by ones that care about attributes
+        // add the changes and transaction to the batched transactions of the observers that care about it.
+        self.attributes.iter().filter_map(|(key, attrs)| {
+            let intersect: AttributeSet = attrs.intersection(changes).cloned().collect();
```

Comment:
Minor performance note: you're collecting the entire cloned intersection inside a loop in order to figure out whether the intersection is empty…

Can we define this as "an observer is called with the full set of changes if any of its attributes match"? I expect most observers will watch only one or a handful, and typically won't care about the details. That allows us to short-circuit here.